stdlib: Mv resource download to `get_local_path` and add `ShadowResource`

BobbyRBruce opened the pull request:
This change decouple's the downloading of a resource from it's data. With this change the `obtain_resource` function returns the `AbstractResource` implementation which contains the data. The resource itself (e.g., the actual disk image, binary, file, etc.) is only downloaded to the host system, if not already present, upon the `get_local_path` call.

`get_local_path` is the function used by gem5 to ultimately load the resource into a simulation, therefore this change ensures we only download resources when they are loaded into a simulation.

This change is not ideal and comes with the following caveats:

1. The `downloader` function is created in `obtain_workload` and passed to the `AbstractResource` implementation for later use. This function comes with the following requirements:
    * The function will download the resource to `local_path`.
    * The function will not re-download the resources if already present
as this function is called _everytime_ `get_local_path` is called.
2. The directories needed to store `local_path` are created in `obtain_workload` regardless. Ergo even if the resource is not used and `get_local_path` is never called these directories are still created.


In keeping with this efficiency `ShadowResource` is introduced to allow the storing of just the resource ID and Version of a resource with additional information only obtained when requested.

BobbyRBruce committed:
stdlib: Move resource download to `get_local_path`

This change decouple's the downloading of a resource from it's data.
With this change the `obtain_resource` function returns the
`AbstractResource` implementation which contains the data. The resource
itself (e.g., the actual disk image, binary, file, etc.) is only
downloaded to the host system, if not already present, upon the
`get_local_path` call.

`get_local_path` is the function used by gem5 to ultimately load the
resource into a simulation, therefore this change ensures we only
download resources when they are loaded into a simulation.

This change is not ideal and comes with the following caveats:

1. The `downloader` function is created in `obtain_workload` and passed
to the `AbstractResource` implementation for later use. This function
comes with the following requirements:
    * The function will download the resource to `local_path`.
    * The function will not re-download the resources if already present
as this function is called _everytime_ `get_local_path` is called.
2. The directories needed to store `local_path` are created in
`obtain_workload` regardless. Ergo even if the resource is not used and
`get_local_path` is never called these directories are still created.

Change-Id: I3f0e9a0099cba946630d719c3d17b7da0bccf74a

powerjg commented:
The approach seems sound to me. 

My one concern is the utilities that we use to download resources. I assume they are unaffected?

Also, it looks like you need to rebase :)

BobbyRBruce committed:
stdlib: Add `ShadowResource`

The purpose of a `ShadowResource` is a resource which only contains the
ID and Version information, not any additional information about the
resource thus avoiding the `obtain_resource` call.

When attributes of the `ShadowResource` are accessed which can only be
obtained via `obtain_resource` the `ShowResource` calls the function and
returns what is required.

This is useful for `Suite` resources which contain several workloads
and resources which may not all be needed when the `Suite` object is
first instantiated.

Change-Id: Icc56261b2c4d74e4079ee66486ddae677bb35cfa

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -750,6 +807,52 @@ def get_input_groups(self) -> Set[str]:
+        """
+        self._params[parameter] = value
+
+    def get_category_name(cls) -> str:
+        return "WorkloadResource"
```
if this is only used for lazy workloads, then instead of `ShadowResource` this class should be something like `Lazy/ShadowWorkloadResource`

BobbyRBruce replied:
I didn't mean to leave in these functions. Removing.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -750,6 +807,52 @@ def get_input_groups(self) -> Set[str]:
+            self._workload = self._obtain_resource_call()
+        return getattr(self._workload, attr)
+
+    def set_parameter(self, parameter: str, value: Any) -> None:
+        """
+        Used to set or override a workload parameter
+
+        :param parameter: The parameter of the function to set.
+        :param value: The value to set to the parameter.
+        """
+        self._params[parameter] = value
```
Why do we need to override this function?

BobbyRBruce replied:
No, that was a mistake. Didn't mean to include it.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -750,6 +807,52 @@ def get_input_groups(self) -> Set[str]:
+        self._workload: Optional[AbstractResource] = None
+        self._obtain_resource_call = obtain_resource_call
+
+    def __getattr__(self, attr):
+        """if getting the id or resource version, we keep the object in the
+        "shdow state" where the `obtain_resource` function has not been called.
+        When more information is needed by calling another attribute, we call
+        the `obtain_resource` function and store the result in the `_workload`.
+        """
+        if attr in {"get_id", "get_resource_version"}:
+            return getattr(super(), attr)
+        if not self._workload:
+            self._workload = self._obtain_resource_call()
+        return getattr(self._workload, attr)
```
Could we put this code in the Resource class?

BobbyRBruce replied:
I think we _could_ but i'm not sure we _should_. It does add a little more complexity and it's only use is in the `Suite` and, maybe, `Workload`s if we want to use it there too (even in this case I don't see much value in it)

powerjg replied:
Ack

powerjg reviewed (commented):
I fear this implementation is now too complicated to maintain easily. However, I'll be happy to hear you out for why the complexity is needed.

Why do we need a new resource class?

BobbyRBruce commented:
> I fear this implementation is now too complicated to maintain easily. However, I'll be happy to hear you out for why the complexity is needed.
> 
> Why do we need a new resource class?

It's the least complex solution available to us. I don't particularly like it but the cost of having to get all workloads in a suite is too high. The only alternative to this would be a massive refactoring of our Client interaction to bundle resource requests.

In this design a Suite's workloads only store the ID and Version until that workload is accessed in any way other than to get the ID/Version. When this happens it is obtained via the `obtain_resource` function.

I think this is superior to #609 as when the Workloads are downloaded is more sensible (i.e., not just the first time the `Suite` is iterated through) and **only** the Workloads needed are ultimately downloaded. With this you can iterate through a suite and see the workloads at a surface level and only download their data when more detailed information is required.
